feat(lua): add support for lua keymaps

shadmansaleh opened the pull request:
Supersedes #13823

This introduces two new functions `vim.keymap.set` & `vim.keymap.del`

differences compared to regular set_keymap:
- remap is used as opposite of noremap. By default it's true for <Plug> keymaps and false for others.
- rhs can be lua function.
- mode can be a list of modes.
- replace_keycode option for lua function expr maps, true by default.
- handles buffer specific keymaps

Examples:
```lua
vim.keymap.set('n', 'asdf', function() print("real lua function") end)
vim.keymap.set({'n', 'v'}, '<leader>lr', vim.lsp.buf.references, {buffer=true})
vim.keymap.set('n', '<leader>w', "<cmd>w<cr>", {silent = true, buffer = 5 })
vim.keymap.set('i', '<Tab>', function()
  return vim.fn.pumvisible() == 1 and "<C-n>" or "<Tab>"
end, {expr = true})
vim.keymap.set('n', '[%', '<Plug>(MatchitNormalMultiBackward)')

vim.keymap.del('n', 'asdf')
vim.keymap.del({'n', 'i', 'v'}, '<leader>w', {buffer = 5 })

clason commented:
I'm sorry, but I have to 👎🏻 on the API (although I of course appreciate the PR itself!)

In my opinion, mimicking the vim script API is an anti-goal; I'd much prefer the actual Neovim API (`nvim_set_keymap`) where the mode is an argument. 

One thing I would love to see in a new API is the possibility to map multiple modes in a single call -- either via a composite `'cnx'` string, or a proper table `{ 'c', 'n', 'x' }`. (The latter is probably easier to reason about.)

I also strongly believe that this is an opportunity to get rid of the `nnoremap` nonsense by turning this into a `remap` option that is false by default.

bfredl commented:
I'd suggest splitting this into two parts in separate PRs, which are relatively independent: (1) the support for luaref:s stored in a mapping. This can be accessed with a `func` option key to `nvim_{buf_}set_keymap`  or something. (2) the new high-level interface for defining mappings in init.lua.

bfredl commented:
> nvim_{buf_}set_keymap is rpc api right ? Does luaref make any sence in that context . For example what can you do with luaref from python/js ?

luaref arg will only work when called from lua, it's the same as any other API func (like nvim_buf_attach, nvim_open_term, etc).

shadmansaleh commented:
@bfredl like #16594 ?

shadmansaleh committed:
feat(lua): add vim.keymap

This introduces two new functions `vim.keymap.set` & `vim.keymap.del`

differences compared to regular set_keymap:
- remap is used as opposite of noremap. By default it's true for <Plug> keymaps and false for others.
- rhs can be lua function.
- mode can be a list of modes.
- replace_keycodes option for lua function expr maps. (Default: true)
- handles buffer specific keymaps

Examples:
```lua
vim.keymap.set('n', 'asdf', function() print("real lua function") end)
vim.keymap.set({'n', 'v'}, '<leader>lr', vim.lsp.buf.references, {buffer=true})
vim.keymap.set('n', '<leader>w', "<cmd>w<cr>", {silent = true, buffer = 5 })
vim.keymap.set('i', '<Tab>', function()
  return vim.fn.pumvisible() == 1 and "<C-n>" or "<Tab>"
end, {expr = true})
vim.keymap.set('n', '[%', '<Plug>(MatchitNormalMultiBackward)')

vim.keymap.del('n', 'asdf')
vim.keymap.del({'n', 'i', 'v'}, '<leader>w', {buffer = 5 })
```

shadmansaleh commented:
Currently this is based on #16594. I'll rebase it on master once
that pr gets merged.

See if this version of keymap api seems better.

This is how possible use cases can look like:
```lua
vim.keymap.add {'n', 'lhs', function() print("real lua function") end, silent = true }
vim.keymap.add {{'n', 'v'}, '<leader>lr', vim.lsp.buf.references, silent = true, buffer=true }
vim.keymap.add {mode = 'n', lhs = '<leader>w', rhs = "<cmd>w<cr>", silent = true, buffer = 5 }
vim.keymap.add {mode = 'i', lhs = '<Tab>', rhs = function()
  return vim.fn.pumvisible() == 1 and "<C-n>" or "<Tab>"
end, silent = true, expr = true, replace_termcodes = true }


vim.keymap.remove {'n', 'lhs'}
vim.keymap.remove {mode = {'n', 'i', 'v'}, lhs = '<leader>w', buffer = 5 }
```

clason commented:
I think I'd prefer 
```lua
vim.keymap.add(mode, lhs, rhs, opts)
```
where
* `mode` is a character or a table of characters (or a string? but then order would matter)
* `lhs` is a string
* `rhs` is a string or (Lua) function
* `opts` is a table with the usual keys (and possibly more in the future), especially `remap = false` (default)

Originally I'd have suggested keeping buffer mappings as a separate function, but the possibility of specifying the buffer directly looks pretty neat.

shadmansaleh commented:
> especially remap = false

having noremap = true as default already does the same . I don't think renaming it to remap is worth it as it's already well known

 > I'd counsel against having multiple different ways to do the same things, and relying on weak typing for positional table arguments strikes me as needlessly brittle -- the mode, lhs, rhs triple is so basic that I don't see the advantage of the flexibility you gain in return?

I have no issue with having them as just positional args . I think tj originally added positional+kwarg to have option for more verbose configuration 🤷‍♂️

Even then I'd like to take only table as input . I really like
```lua
add{, , , x = t, y = f}
```
vs
```
add(, , , {x = t, y = f})
```

clason commented:
This is just my opinion, but I want to clarify that it is a strong one (I won't debate this further):

> having noremap = true as default already does the same . I don't think renaming it to remap is worth it as it's already well known

I very much do, since the double negative is just a needless mental hurdle at this point. (How many people treat `nnoremap` as a magic incantation?)

> Even then I'd like to take a table as input . I really like [...]

I _very, very much_ prefer the latter  ;)

shadmansaleh commented:
> I very much do, since the double negative is just a needless mental hurdle at this point. (How many people treat nnoremap as a magic incantation?)

Unless someone's making a <Plug> map one most likely won't need to touch the noremap flag in this setup . Recursive keymaps are already a niche topic and it's more likely someone to stumble upon remap's explanation if they are looking for noremap rather then remap . Because noremap is already established term renaming it now can only add to the confusion then resolve it in my opinion.

> I very, very much prefer the latter ;) It's also a more standard pattern across the API, and likely more robust.

vim.keymap is just lua syntactic suger over underling nvim_set_keymap . rpc api has to use language agnostic  interface but that doesn't apply to lua config wrappers in my opinion (vim.wo[3].nu ?) . Otherwise what's the point of vim.keymap / vim.opt vim.api..... functions can already do everything .

bfredl commented:
> Recursive keymaps are already a niche topic and it's more likely someone to stumble upon remap's explanation if they are looking for noremap rather then remap .

They are not _that_ niche as you still need them for `<Plug>` mappings which still are gonna be a thing. And IMO `remap=true` is both more readable and succinct (say what you mean) than `noremap=false` which not even was a modifier originally, but the full name of the map command. Stumbling one way or the other can be helped with proper tags and cross-references in the docs.

shadmansaleh commented:
ok two wins over one :P

clason commented:
> ok two wins over one :P

Nah, just bfredl wins over (nearly) everyone ;)

(He didn't comment on the function signature, though, so maybe wait for more feedback from him and others?)

gpanders commented:
> > ok two wins over one :P
> 
> Nah, just bfredl wins over (nearly) everyone ;)
> 
> (He didn't comment on the function signature, though, so maybe wait for more feedback from him and others?)

I agree with @clason on the function signature as well. I also prefer

vim.keymap.add(,,,{x = 1, y = 2})

Using a table to mix positional and keyword args feels “too clever” to me. If individual users or plugins want to do it then more power to them, but for the officially sanctioned Nvim API I’m in favor of keeping it simple.

Although while we’re bikeshedding over names, why not `vim.keymap.set` and `vim.keymap.del` to match the API functions?

shadmansaleh commented:
Ok now since api part of this is merged . We can debate on what lua interface will be the best . Personally I feel like if vim.keymap is just copying nvim_set_keymap it doesn't need to exist . I think if vim.keymap is to be it should be to make configuration easier instead of what's more programatic vim.api...... already covers that :P .

And about the naming part (add/set|remove/del) I don't have a strong opinion here . Whatever the majority chooses :)

clason commented:
I don't know, the ability to
1. map multiple modes
2. map Lua functions without callback shenanigans (bonus points if it takes `foo(bar)` without manual wrapping) and
3. get rid of the `noremap=false` nonsense

is pretty compelling for me already ;)

Also, as a Lua API, we are free to add and change things later (within reason), which we can't for the Neovim API.

max397574 commented:
As far as I can tell this summarizes what this pr allows:
```lua
 vim.keymap.add{'n', 'lhs', function() print("real lua function") end, silent = true } -- function as rhs

vim.keymap.add{{'n', 'v'}, '<leader>lr', vim.lsp.buf.references, silent = true, buffer=true } -- lua command as rhs, multiple modes

vim.keymap.add{'n', '<leader>w', "<cmd>w<cr>", silent = true, buffer = 5} -- still allow "normal" mappings

vim.keymap.add{'i', '<Tab>', function() -- function with replace-termcodes
  return vim.fn.pumvisible() == 1 and "<C-n>" or "<Tab>"
end, silent = true, expr = true, replace_termcodes = true }
```

For me as a plugin dev it would be really interesting to be able to do something like this:
```lua
local function reload()
  clear()
  load_plugin()
end

vim.keymap.add{'n', '<leader>r', reload(), silent=true}
```
 I don't care if it's `reload()` or `function() reload() end`
 I just want to be able to use local functions in mapping.
 before you always had to do
 ```lua
local my_module={}
function my_module.reload()
  clear()
  load_plugin()
end

vim.api.nvim_set_keymap('n', '<leader>r', '<cmd>lua require"my_module".reload()', {silent=true})

return my_module
```
and this is in my opinion was too complicated for such a simple thing.
Also it's something that you really often need in a plugin to create mappings for example for a menu or a buffer with special functions.

I don't know if this already is possible with this pull request. If not it would be really great if you could implement it.

lewis6991 commented:
> For me as a plugin dev it would be really interesting to be able to do something like this:
>
> ```lua
> local function reload()
>   clear()
>   load_plugin()
> end
>
> vim.keymap.add{'n', '<leader>r', reload(), silent=true}
> ```
 You would have to drop the parenthesis to pass a function ref as opposed to calling the function:

```lua
vim.keymap.add{'n', '<leader>r', reload, silent=true}
```

shadmansaleh commented:
> vim.keymap.add{{'n', 'v'}, '<leader>lr', vim.lsp.buf.references, silent = true, buffer=true } -- lua command as rhs, multiple modes

@max397574  This is not lua command . It is a lua function too . It's directly passing the function reference.

As lewis have said you can provide your function reference too 

For example:
```lua
local function reload()
  clear()
  load_plugin()
end

vim.keymap.add{'n', '<leader>r', reload, silent=true}
```

Notice it doesn't have parans after reload as you aren't calling reload rather passing reload function as parameter.

> Also it's something that you really often need in a plugin to create mappings for example for a menu or a buffer with special functions.

I'm not sure what you mean by special functions can you elaborate on that ?

max397574 commented:
For example I've been working on a help file viewer in a float. And this has the "special function" (perhaps this was the wrong word) that I can use J/K no navigate to next/previous files.
So I had to do something like
```lua
local float_help={}
float_help.next=function()
...
end

float_help.prev=function()
...
end

vim.api.nvim_buf_set_keymap(buf, "n", "J", "<cmd>lua require'float_help'.next()<CR>", {noremap=true, silent=true})
-- same for prev

return float_help()
```
and now I could just
```lua
next()
...
prev()...

vim.keymap.add{'n', 'J', next, silent=true, buffer=5} -- iirc that's the syntax for buffers but not sure
```

shadmansaleh commented:
> and now I could just
>```lua
>local next()
>...
>local prev()...
>
>vim.keymap.add{'n', 'J', next, silent=true, buffer=buf} -- iirc that's the syntax for buffers but not sure
>```

Yes you'll be able to do that.

tjdevries commented:
I will try to review this within the next week.

clason commented:
I've played around with the new luaref callbacks and noted that a naive translation is significantly more expensive: Going from
```lua
vim.api.nvim_set_keymap('n', 'f', '<cmd>lua require"foo".bar()<cr>', {})
```
to
```lua
vim.api.nvim_set_keymap('n', 'f', '', {callback =  require"foo".bar})
```
is about 50x slower since the `require'foo'` is actually executed. This can be mitigated by instead doing 
```lua
vim.api.nvim_set_keymap('n', 'f', '', {callback =  function() require"foo".bar() end})
```
This also affects things like `vim.lsp.buf.hover`, albeit to a smaller extent.

All this is a long-winded way of saying that it would be nice if `vim.keymap.set` could (optionally) handle this wrapping for me -- I believe that the vast majority of users would want that?

gpanders commented:
@clason Unfortunately I don’t think what you’re asking for is possible without some kind of macro system which Lua sadly lacks.

Once we rewrite Neovim in Fennel we could do that trivially :)

clason commented:
Well, you can do similar things as in https://github.com/nvim-lualine/lualine.nvim : allow strings to be executed as `luaexpr` (an option) via `loadstring`?

shadmansaleh commented:
@clason I'm not sure if adding that to core will be a good idea.
But you can always use somthing like this as part of your config.
```lua
local l = loadstring
vim.api.nvim_set_keymap('n', 'f', '', {callback = l"require'foo'.bar()"})
```

clason commented:
I'm not sure either, but I wanted to raise the point here and now, while we're discussing the "convenience" API (and I do think this would be convenient ;))

gegoune commented:
I think that if `require` based calls execute at map registration taking up time it should at least be mentioned in documentation. I am excited about this PR and would be bitten by this issue without realising what's caused slowdowns in my nvim startup as this is exactly what I planned to do for quite a few mappings.

clason commented:
The biggest benefit of the native mappings will in any case be for plugins (or core) that directly map local or in-module functions, where this will not be relevant.

muniter commented:
A workaround would be to provide some simple "lazy require" in core. 

```lua
vim.api.nvim_set_keymap('n', 'f', '', {callback = lazy_require"foo".bar})
```

Something like Teej example plugin https://github.com/tjdevries/lazy.nvim

It isn't too magic and allows keeping the convenience we all want.

gpanders commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,127 @@
+--- Add a new keymap.
+--- Examples:
+--- <pre>
+---   -- You can make keymaps worth lua functions
```
"Lua" is a proper noun, so in documentation it should always be capitalized.

```suggestion
---   -- You can make keymaps worth Lua functions
```

gpanders commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,127 @@
+---   -- You can map a key for specific buffer
+---   vim.keymap.set('n', '<leader>w', "<cmd>w<cr>", { silent = true, buffer = 5 })
+---
+---   -- Also expr keymaps too. Notice the replace_termcodes when making lua expr maps.
```
```suggestion
---   -- Also expr keymaps too. Notice the replace_termcodes when making Lua expr maps.
```

gpanders commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,127 @@
+---   -- While setting <Plug> mapping remember to set remap to true
+---   vim.keymap.set('n', '[%', '<Plug>(MatchitNormalMultiBackward)', {silent = true, remap = true })
+--- </pre>
+--@param mode (string|table):   Same mode short names as |nvim_set_keymap|.
```
```suggestion
---@param mode string|table   Same mode short names as |nvim_set_keymap()|.
```

gpanders commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,127 @@
+--- </pre>
+--@param mode (string|table):   Same mode short names as |nvim_set_keymap|.
+---                             Can also be list of modes to create keymap on multiple modes.
+--@param lhs (string):          Left-hand-side |{lhs}| of the mapping.
```
```suggestion
---@param lhs string          Left-hand-side |{lhs}| of the mapping.
```

gpanders commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,127 @@
+--@param mode (string|table):   Same mode short names as |nvim_set_keymap|.
+---                             Can also be list of modes to create keymap on multiple modes.
+--@param lhs (string):          Left-hand-side |{lhs}| of the mapping.
+--@param rhs (string|function): Right-hand-side |{rhs}| of the mapping. Can also be a lua function.
```
```suggestion
---@param rhs string|function Right-hand-side |{rhs}| of the mapping. Can also be a Lua function.
```

gpanders commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,127 @@
+--@param lhs (string):          Left-hand-side |{lhs}| of the mapping.
+--@param rhs (string|function): Right-hand-side |{rhs}| of the mapping. Can also be a lua function.
+--
+--@param opts (table): A table of |:map-arguments| such as "silent". See |nvim_set_keymap()|
+---       along with "replace_termcodes". When replace_termcodes is true with
+---       lua function expr maps |nvim_replace_termcodes| is applied to return
+---       value of mapped function.
+---       One difference compared to |nvim_set_keymap| is here "remap" used as option
+---       instead of |noremap| . It's meaning is opposite of |noremap| and it's false by default.
+--@see |nvim_set_keymap|
+---
```
```suggestion
---@param opts table A table of |:map-arguments| such as "silent". In addition to the options
---                  listed in |nvim_set_keymap()|, this table also accepts the following keys:
---                  - replace_termcodes: (boolean, default false) when "true",
---                  |nvim_replace_termcodes()| is applied to the result of Lua expr maps.
---                  - remap: (boolean, default false) Make the mapping recursive. This is the
---                  inverse of the "noremap" option from |nvim_set_keymap()|.

---@see |nvim_set_keymap()|
```

gpanders commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,127 @@
+--@see |nvim_set_keymap|
+---
+function keymap.set(modes, lhs, rhs, opts)
+  opts = opts or {}
+
+  vim.validate {
+    mode = {modes, {'s', 't'}},
+    lhs = {lhs, 's'},
+    rhs = {rhs, {'s', 'f'}},
+    opts = {opts, 't'}
+  }
```
Putting `vim.validate` at the very top of the function is a bit more idiomatic.

```suggestion
  vim.validate {
    mode = {modes, {'s', 't'}},
    lhs = {lhs, 's'},
    rhs = {rhs, {'s', 'f'}},
    opts = {opts, 't', true}
  }
  
  opts = opts or {}
```

gpanders commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,127 @@
+---
+---   vim.keymap.del({'n', 'i', 'v'}, '<leader>w', { buffer = 5 })
+--- </pre>
+--@param opts (table): A table optional arguments:
```
```suggestion
---@param opts table A table of optional arguments:
```

gpanders commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,127 @@
+--@param opts (table): A table optional arguments:
+---     - buffer: Can be boolean or bufnr . Needed when removing buffer
+---               local keymaps. buffer true or 0 refers to current buffer.
+--@see |vim.keymap.set|
```
```suggestion
---@see |vim.keymap.set()|
```

gpanders commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,127 @@
+--@see |vim.keymap.set|
+---
+function keymap.del(modes, lhs, opts)
+  opts = opts or {}
+
+  vim.validate {
+    mode = {modes, {'s', 't'}},
+    lhs = {lhs, 's'},
+    opts = {opts, 't'}
+  }
```
Same thing here, make `vim.validate` the first thing and make the `opts` argument optional, and put `opts = opts or {}` below `vim.validate`

gpanders commented on runtime/lua/vim/shared.lua:
```diff
@@ -526,13 +526,23 @@ end
 --- </pre>
 ---
----@param opt Map of parameter names to validations. Each key is a parameter
+--- If multiple types are valid they can be given as a list.
+--- <pre>
+---  vim.validate{arg1={{'foo'}, {'table', 'string'}}, arg2={'foo', {'table', 'string'}}}
+---     => NOP (success)
+---
+---  vim.validate{arg1={1, {'string', table'}}}
+---     => error('arg1: expected string|table, got number')
+---
+--- </pre>
+---
+---@param opt table of parameter names to validations. Each key is a parameter
```
Should these changes be removed in light of #16864?

gpanders commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,127 @@
+---     return vim.fn.pumvisible() == 1 and "<C-n>" or "<Tab>"
+---   end, { silent = true, expr = true, replace_termcodes = true })
+---   -- While setting <Plug> mapping remember to set remap to true
+---   vim.keymap.set('n', '[%', '<Plug>(MatchitNormalMultiBackward)', {silent = true, remap = true })
+--- </pre>
```
I think it's worth mentioning an example of using `require` from a plugin. As @clason mentioned this will be a common use case and it's likely many people will just attempt to use `require("foo").some_func` which of course will load the `foo` plugin when the keymap is defined (which isn't what people want). Having a canonical example of using `function() require("foo").some_func() end` would be good.

clason replied:
Agreed; much of the issue is that we simply haven't worked out best practices around these new Lua variants of well-established vimscript patterns. (@nanotee to the rescue ;))

I'll have to play around whether `loadstring` or anonymous functions are the more convenient pattern (or what the exact trade-offs are for each); in any case, both are supported in the current API, so any further improvement can -- and should! -- come later.

(The _real_ drawback with the `function()` thing is that `stylua` insists on making it multiline...)

tjdevries replied:
you could also do `require_on_call` (https://github.com/tjdevries/lazy.nvim/blob/master/lua/lazy.lua#L49-L72) but people will have different preferences on how to do so I think.

clason replied:
I think "lazy require" is a good idea but independent from this PR (as in, I don't see how the implementation or API would change if that were available), so it's better to discuss this as part of the general "require goes *brrrr*" discussion around caching etc., for example in https://github.com/neovim/neovim/pull/15436.

Mentioning that Lua functions are actually accessed during `set` (so `require` is executed) and mentioning the `function` wrapper workaround is enough here. I'd prefer it if this PR got merged early and then iterated on during the 0.7 development cycle.

gpanders reviewed (changes requested):
The code LGTM, I just had feedback on the docs.

gpanders commented on runtime/doc/lua.txt:
```diff
@@ -1778,4 +1778,77 @@ select({items}, {opts}, {on_choice})                         *vim.ui.select()*
+                            |nvim_set_keymap()|. Can also be list of modes to
+                            create keymap on multiple modes.
+                    {lhs}   string Left-hand-side |{lhs}| of the mapping.
+                    {rhs}   string|function Right-hand-side |{rhs}| of the
```
Hmm looks like we can't use `|` as the type separator here since it's interpreted as a tag by the help syntax. I'm not sure what to use instead, any ideas?

shadmansaleh replied:
How about '/' ? Though I think emmy ;ua notation uses '|' .

shadmansaleh replied:
Huh . It's actually alright . Open the file in neovim it gets displayed correctly . Githubs highlighting is broken.

gpanders commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,136 @@
+--- Add a new keymap.
+--- Examples:
+--- <pre>
+---   -- You can make keymaps worth Lua functions
```
What does "worth" mean in this context? Maybe re-word this to simply
```suggestion
---   -- You can use Lua functions
```

shadmansaleh replied:
typo *with 🤦‍♂️

gpanders commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,136 @@
+---   vim.keymap.del({'n', 'i', 'v'}, '<leader>w', { buffer = 5 })
+--- </pre>
+---@param opts table A table of optional arguments:
+---     - buffer: Can be boolean or bufnr . Needed when removing buffer
+---               local keymaps. buffer true or 0 refers to current buffer.
```
```suggestion
---     - buffer: (number or boolean) Remove a keymap from the given buffer. When "true" or 0, use the current buffer.
```

tjdevries commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,136 @@
+    opts.replace_termcodes = nil
+  end
+
+  -- remap defaults to false
```
is it crazy to check something like `<Plug>` in the rhs if it's a string and then turn remap on by default?

shadmansaleh replied:
If it's not considered too magical for core, I have no issue with it 😆 . 

Should we be changing default value of an option based on other option ? I can definitely see this being more convenient . The whole remap noremap thing is a mess for any new comer .

clason replied:
Could also leave it for a separate follow-up PR? I don't think adding that would be a breaking change (as opposed to removing it).

shadmansaleh replied:
It's 3 lines of code we can just decide whether we want it or not . `<Plug>` key can't be used without remap . I doubt anyone would be mad even if the internal `:map` mechanics treated `<Plug>` mappings to always be remap .

tjdevries commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,136 @@
+    opts.buffer = nil
+  end
+
+  if opts.replace_termcodes ~= nil then
```
My preference is this should be on by default, but I haven't read all the comments yet, so perhaps this has already been discussed.

shadmansaleh replied:
We haven't discussed it here . Initially I had [asked](https://github.com/neovim/neovim/pull/16594#discussion_r766309821) bfredl about applying replace termcode always on return of expr maps . But then it was declined . But I think making this default to true can be done . If someone doesn't want it they can set it to false . As far as I can see expr mapping almost always needs replace_termcodes

gpanders replied:
+1 for making it the default from me too

clason replied:
>  Initially I had asked bfredl about applying replace termcode always on return of expr maps . But then it was declined .

This was for the underlying Neovim API; I believe the Lua API is allowed to be (slightly) more magic. @bfredl?

gpanders commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,136 @@
+--
+---@param opts table A table of |:map-arguments| such as "silent". In addition to the options
+---                  listed in |nvim_set_keymap()|, this table also accepts the following keys:
+---                  - replace_termcodes: (boolean, default false) when "true",
```
```suggestion
---                  - replace_termcodes: (boolean, default false) When "true",
```

mjlbach commented on runtime/doc/lua.txt:
```diff
@@ -1778,4 +1778,77 @@ select({items}, {opts}, {on_choice})                         *vim.ui.select()*
+                    vim.keymap.set('n', 'asdf', require('jkl').my_fun)
+<
+
+                `require('jkl)` get's evaluated durimg this call. If you want to avoid this
```
get's -> gets

mvllow replied:
```diff
- `require('jkl)`
+ `require('jkl')`
```

shadmansaleh commented:
I think we should have remap enabled on `<Plug>` keymaps . As there's really no meaning of `<Plug>` mappings without the remap, the mapping just won't work . I don't think anyone is intentionally making a `<Plug>` mapping without remap . So makes sense to remap `<Plug>` mappings by default . Also this way the user almost never will need to touch the `remap`, that's quite convenient IMO and avoid lots of confusion .

clason commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,141 @@
+--- Add a new keymap.
+--- Examples:
+--- <pre>
+---   -- You can make keymaps with Lua functions
```
```suggestion
---   -- Can map to Lua functions
```

clason commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,141 @@
+---   -- Can use it to map multiple modes
+---   vim.keymap.set({'n', 'v'}, '<leader>lr', vim.lsp.buf.references, { silent = true, buffer=true })
+---
+---   -- You can map a key for specific buffer
```
```suggestion
---   -- Can add map for specific buffer
```

clason commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,141 @@
+---   vim.keymap.set('n', '[%', '<Plug>(MatchitNormalMultiBackward)', {silent = true})
+--- </pre>
+---
+--- A pit fall of this is if you want to make keymap like:
```
```suggestion
--- A pitfall of this is if you want to make keymap like
```

clason commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,141 @@
+---    vim.keymap.set('n', 'asdf', require('jkl').my_fun)
+--- </pre>
+---
+--- require('jkl') gets evaluated durimg this call. If you want to
```
```suggestion
--- the `require('jkl')` gets evaluated during this call in order to access the function. If you want to
```

clason commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,141 @@
+--- </pre>
+---
+---@param mode string|table   Same mode short names as |nvim_set_keymap()|.
+---                            Can also be list of modes to create keymap on multiple modes.
```
Can we be consistent here about "map" | "keymap" | "mapping"?

clason replied:
Can also add a link to `|mapping|`.

clason commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,141 @@
+---
+---@param mode string|table   Same mode short names as |nvim_set_keymap()|.
+---                            Can also be list of modes to create keymap on multiple modes.
+---@param lhs string          Left-hand-side |{lhs}| of the mapping.
+---@param rhs string|function  Right-hand-side |{rhs}| of the mapping. Can also be a Lua function.
```
```suggestion
---@param lhs string          Left-hand side |{lhs}| of the mapping.
---@param rhs string|function  Right-hand side |{rhs}| of the mapping. Can also be a Lua function.
```

clason commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,141 @@
+--
+---@param opts table A table of |:map-arguments| such as "silent". In addition to the options
+---                  listed in |nvim_set_keymap()|, this table also accepts the following keys:
+---                  - replace_termcodes: (boolean, default true) When "true",
+---                  |nvim_replace_termcodes()| is applied to the result of Lua expr maps.
```
```suggestion
---                  - replace_termcodes: (boolean, default true) When both this and `expr` are `true`,
---                  |nvim_replace_termcodes()| is applied to the result of Lua expr maps.
```

clason commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,141 @@
+---                  |nvim_replace_termcodes()| is applied to the result of Lua expr maps.
+---                  - remap: (boolean) Make the mapping recursive. This is the
+---                  inverse of the "noremap" option from |nvim_set_keymap()|.
+---                  By default it's true for <Plug> mappings and false for non <Plug> mappings.
```
```suggestion
---                  Default `false` unless `lhs` is a string starting with `<plug>` (case-insensitive).
```

gpanders commented on runtime/doc/lua.txt:
```diff
@@ -1811,21 +1811,22 @@ set({mode}, {lhs}, {rhs}, {opts})                           *vim.keymap.set()*
 
-                `require('jkl)` get's evaluated durimg this call. If you want to avoid this
-                cost at startup you can wrap it in a function, for example: >
+                require('jkl') gets evaluated durimg this call. If you want to
```
```suggestion
                require('jkl') gets evaluated during this call. If you want to
```

shadmansaleh replied:
Wait what I just fixed that 🤷‍♂️

shadmansaleh replied:
It's already been fixed . You're looking at older diff.

tjdevries commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,138 @@
+  local is_rhs_lua = type(rhs) == "function"
+  mode = type(mode) == 'string' and {mode} or mode
+
+  if opts.buffer ~= nil then
+    buffer = opts.buffer == true and 0 or opts.buffer
+    opts.buffer = nil
+  end
```
This might be nicer to write it like:

```lua
if opts.buffer then
  local buffer = opts.buffer == true and 0 or opts.buffer
  -- do nvim_buf_set_keymap ...
else 
  -- do nvim_set_keymap ...
end
```

Not sure if that makes sense though.

shadmansaleh replied:
This will leave `bufnr` or `true` in opts.buffer . And make nvim_set_keymap crash because of unknown key.

tjdevries commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,138 @@
+---                  inverse of the "noremap" option from |nvim_set_keymap()|.
+---                  Default `true` if `lhs` is a string starting with `<plug>` (case-insensitive), `false` otherwise.
+---@see |nvim_set_keymap()|
+function keymap.set(mode, lhs, rhs, opts)
```
opts should be copied here -- since we modify it in place throughout the function. This would lead to really surprising things happening for people if they use the same `opts` table for multiple calls.

shadmansaleh replied:
Good idea.

tjdevries commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,138 @@
+    opts.buffer = nil
+  end
+
+  if is_rhs_lua and opts.expr and opts.replace_termcodes ~= false then
```
If you just pass something like "<Tab>" instead of a function, does that also need termcodes replaced or not?

(I'm not sure if this should be guarded by `is_rhs_lua` is sort of the question I'm asking)

shadmansaleh replied:
> If you just pass something like "" instead of a function, does that also need termcodes replaced or not?

No you don't

tjdevries replied:
cool, ty

shadmansaleh commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,138 @@
+
+  opts = opts or {}
+  local buffer = false
+  local is_rhs_lua = type(rhs) == "function"
```
How about `is_rhs_luaref` ? rhs might just be a call to viml function too in which case lua specific modification we make isn't valid .

tjdevries replied:
fine by me

MurdeRM3L0DY commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,138 @@
+  if is_rhs_lua and opts.expr and opts.replace_termcodes ~= false then
+    local user_rhs = rhs
+    rhs = function ()
+      return vim.api.nvim_replace_termcodes(user_rhs(), true, true, true)
```
Hi. May I recommend `vim.api.nvim_replace_termcodes(vim.api.nvim_eval(user_rhs()), true, true, true)` so mappings like:
```lua
vim.keymap.set('<tab>', function()
  return "luasnip#expand_or_jumpable() ? '<Plug>luasnip-expand-or-jump' : '<tab>'"
end, { expr = true })
```
would work.

clason replied:
This specific snippet is not a compelling example since it can be completely translated to a Lua (expr) mapping.

shadmansaleh replied:
I'd say no because then lua functions directly returning keycodes will not work .

clason replied:
or even 
```lua
vim.keymap.set('i', '<tab>', function()
  if require"luasnip".expand_or_jumpable() then
    require'luasnip'.expand_or_jump()
  else
    return "<tab>"
  end
end, { expr = true })
```

clason replied:
(which I assume LuaSnip will eventually implement themselves)

MurdeRM3L0DY replied:
> No because then lua functions directly returning keycodes will not work .
> 
> Also I can't see why you'd be wraping that in `function() end` ? Why not just use
> 
> ```lua
> vim.keymap.set('i', '<tab>', "luasnip#expand_or_jumpable() ? '<Plug>luasnip-expand-or-jump' : '<tab>'", { expr = true })
> ```

True, it wouldn't be reasonable to wrap such a mapping in a lua function.

MurdeRM3L0DY replied:
> (which I assume LuaSnip will eventually implement themselves)

Yes, I use a similar mapping.

MurdeRM3L0DY replied:
> or even
> 
> ```lua
> vim.keymap.set('i', '<tab>', function()
>   if require"luasnip".expand_or_jumpable() then
>     require'luasnip'.expand_or_jump()
>   else
>     return "<tab>"
>   end
> end, { expr = true })
> ```

Yes, I use a similar mapping.

dstanberry replied:
> or even
> 
> ```lua
> vim.keymap.set('i', '<tab>', function()
>   if require"luasnip".expand_or_jumpable() then
>     require'luasnip'.expand_or_jump()
>   else
>     return "<tab>"
>   end
> end, { expr = true })
> ```

Hopefully you will forgive the off-topic question, but does this mapping work for you?
I've been trying unsuccessfully so far to port my luasnip keybinds but no luck so far.. the above example produces a `Failed to save undo information` error.

bfredl commented on runtime/lua/vim/keymap.lua:
```diff
@@ -0,0 +1,135 @@
+--
+---@param opts table A table of |:map-arguments| such as "silent". In addition to the options
+---                  listed in |nvim_set_keymap()|, this table also accepts the following keys:
+---                  - replace_termcodes: (boolean, default true) When both this and expr is "true",
```
new api:s should use "keycodes" instead of "termcodes" as these are not specific to the terminal.

tjdevries commented:
@bfredl what do you think? Any other changes you'd want to make? Or anything you think is too magical?

bfredl commented:
@tjdevries I think it looks good. setting remap automatically on `<Plug>` _is_ a bit magic, but my hunch is that it is precisely the right amount of magic ("lagom") in this case :].

tjdevries commented:
Yeah, the `<plug>` business is magic, but I've been bitten a few times by _not_ doing the remap, so I imagine newer people would be very confused as well (honestly, I wish anything with <plug> as a key just automatically went to remap. There is no reason to even have <plug> mappings otherwise).

RaZ0rr-Two commented:
In the docs for `opts` for `vim.keymap.set,` there is `remap`. Why not make it `recursive` instead when taking into consideration readability ?
Imo having keys such as `remap` and `noremap` are the same. They, more or less, take the same time to understand and apply. `remap` also brings in the confusion of **re-map** as in **mapping again**. So, I think `recursive` makes more change than changing `noremap` to `remap` ( which I don't see as bringing in any improvement :smile: )

tjdevries reviewed (approved):
I think this works for me, but I'll defer to @bfredl on anything else that should be done here.

(re: `recursive` I don't know if I like using that because AFAIK there isn't any usage of the word in the help docs to say what that would do. I think making excellent defaults is good (and even sometimes "magic" defaults) but changing the terminology is a much bigger step because now there are two ways to describe mappings in the docs, and I don't like two words to describe the same thing if we can help it)

nyngwang commented:
How to update this line please:

```
vim.api.nvim_set_keymap('v', '<M-?>', ':Commentary<CR>gv', NOREF_NOERR_TRUNC)
```

I want to exit the visual mode to normal mode and go back in again.

clason commented:
Please ask usage questions on https://neovim.discourse.group, not on old PRs!

dlesnoff commented:
There is one thing unclear to me while reading this PR and the docs.
How does remap without specified mode translate into this API ?
Let us say we had this remap in our Vim config:
```vim
noremap j r
noremap r j
```
This means we switch the "j" and "r" keys in most of the modes : Normal, Visual, Select, Operator-pending (but none of Insertion, Command-Line, Terminal and Lang-Arg).
I understand that we have to rewrite it to:
```lua
vim.keymap.set({'n','v','s','o'}, "j", "r")
vim.keymap.set({'n','v','s','o'}, "r", "j")
```
Is there no other mode simplification ?
Can we somehow set a shortcut for this set of mode ? Like 'a' (for 'almost all') or 'g' for global ?

max397574 commented:
> , or empty string for |:map|.

dlesnoff commented:
Sorry. Just saw this:
> Please ask usage questions on https://neovim.discourse.group/, not on old PRs!
and I skipped the `nvim_set_keymap` help.
Thanks !
I regret the documentation to not  be more explicit: "or empty string for the modes associated to the vim command map which are {'n','v','s','o'}.".
